### ISSUE TITLE
Calculating table columns width like in HTML4

### ISSUE THREAD

**Issue by Intelligent2013** [metanorma/mn-native-pdf#354]
Source: https://github.com/metanorma/iso-10303-detached-docs/issues/104#issuecomment-1103340628

> Instead of using proportional width, I would say if the column proportion exceeds a certain threshold, why don't we implement a proper column adjustment algorithm like in [HTML4](https://www.w3.org/TR/REC-html40/appendix/notes.html#h-B.5.2):
https://stackoverflow.com/questions/22206825/algorithm-for-calculating-variable-column-widths-for-set-table-width

>> The minimum and maximum cell widths are then used to determine the corresponding minimum and maximum widths for the columns. These in turn, are used to find the minimum and maximum width for the table. Note that cells can contain nested tables, but this doesn't complicate the code significantly. The next step is to assign column widths according to the available space (i.e., the space between the current left and right margins).
>> 
>> For cells that span multiple columns, a simple approach consists of apportioning the min/max widths evenly to each of the constituent columns. A slightly more complex approach is to use the min/max widths of unspanned cells to weight how spanned widths are apportioned. Experiments suggest that a blend of the two approaches gives good results for a wide range of tables.
>> 
>> The table borders and intercell margins need to be included in assigning column widths. There are three cases:
>> 
>> * The minimum table width is equal to or wider than the available space. In this case, assign the minimum widths and allow the user to scroll horizontally. For conversion to braille, it will be necessary to replace the cells by references to notes containing their full content. By convention these appear before the table.
>> * The maximum table width fits within the available space. In this case, set the columns to their maximum widths.
>> * The maximum width of the table is greater than the available space, but the minimum table width is smaller. In this case, find the difference between the available space and the minimum table width, lets call it W. Lets also call D the difference between maximum and minimum width of the table.
>>   * For each column, let d be the difference between maximum and minimum width of that column. Now set the column's width to the minimum width plus d times W over D. This makes columns with large differences between minimum and maximum widths wider than columns with smaller differences.

**Comment by Intelligent2013** [comment]
> The minimum and maximum cell widths are then used to determine the corresponding minimum and maximum widths for the columns. 

The term 'width' in HTML specification isn't equal length of text in chars. The 'width' depend on font-size, font-family, etc. In PDF we can't determine 'width' of text without multiple Intermediate Format generation, therefore we need to generate it. It'll increase total processing time. For simplicity let's consider that font is monospaced. May be later I'll think up how to optimize calculation via IF.

> The minimum table width is equal to or wider than the available space. In this case, assign the minimum widths and allow the user to scroll horizontally.

This case in non-acceptable to the PDF. Therefore the current (old) algorithm  will be applied here.

**Comment by Intelligent2013** [comment]
From: https://github.com/metanorma/iso-10303-detached-docs/issues/104#issuecomment-1115208020

> We also need a way to split columns if the total width of the two longest words in the columns exceed page width + margin.

**Comment by Intelligent2013** [comment]
Simple algorithm (based on 'width in chars') realized, but there are a few bad cases:
![image](https://user-images.githubusercontent.com/5559926/166493233-b7d42fc8-417c-476c-a11b-baf861b870ea.png)

![image](https://user-images.githubusercontent.com/5559926/166493381-494fd5e2-897c-442c-bd49-951d756f3cb0.png)


Need to fine tune.

**Comment by Intelligent2013** [comment]
Fixed:
- data from mathml tags `@open` and `@close` includes in total string length now:
![image](https://user-images.githubusercontent.com/5559926/166506396-0253e48c-704a-43af-82af-589c150a3a87.png)

- tags 'br' replaced by 'paragraphs' before algorithm applying;
![image](https://user-images.githubusercontent.com/5559926/166506505-adbff686-4f43-42d0-a14c-747f76db0107.png)

**Comment by Intelligent2013** [comment]
Extra space removed:
![image](https://user-images.githubusercontent.com/5559926/166525878-a33456f4-58d4-4a17-aa20-4a5a685d49fd.png)

**Comment by Intelligent2013** [comment]
Another issue found:
![image](https://user-images.githubusercontent.com/5559926/166551691-6536b436-4c57-4d40-b9bc-7d13f30182ce.png)
![image](https://user-images.githubusercontent.com/5559926/166551985-7dec3f8b-eacb-4bfe-a9fa-c073fbc08582.png)

Enough space for 3 small letters, but here is all capital letters.

**Comment by Intelligent2013** [comment]
Fixed: 
![image](https://user-images.githubusercontent.com/5559926/166561719-cf99f599-9cbb-41a6-9ab2-983dabb2e8a1.png)

![image](https://user-images.githubusercontent.com/5559926/166561780-96be2b3a-aa38-4218-929c-e3d4f1f6ce30.png)

**Comment by Intelligent2013** [comment]
I've checked ISO 108 and 110 PDF and didn't find any issues in table column's width.

Next development phase is IF generation for each table cell and calculate the _real_ width in pixel (instead of char length). It will take more time.

**Comment by Intelligent2013** [comment]
Via Apache FOP Intermediate Format (IF) we need to know these information for each table cell:
- maximum width
- widest element (word, image, etc.)

Some thoughts:
- convert table into simple table without spanned column and rows (made in the xslt already)
- assign unique id for each cell: `tableId_cell_rowNumber_colNumber`.
- turn-off word wrap feature (or create wide paper size (300mm enough), in this case the hair space at the end should be placed in a separate right table cell, to determine max horizontal length)
- render each table cell in a separate lines in a few forms:
   - full content - to determine maximum width (with id `tableId_cell_rowNumber_colNumber` + `full`)
   - each word separately - to determine widest element (width id ``tableId_cell_rowNumber_colNumber` + `word`)
- to determine right margin of each data item, it should starts with new line and ends with hair space with unique id for this data.
- determine available horizontal space for each concrete table
- the algorithm should consider the paper orientation - portrait or landscape (maximum available width).

**Comment by Intelligent2013** [comment]
First result of the accurate width calculation based on IF rendering:

![image](https://user-images.githubusercontent.com/5559926/167911274-e5c5b748-0ad0-434b-97ab-c09d3fbbf198.png)

Debug info:
```xml
<table_width>
	<table width_max="869616" width_min="271810"/>
</table_width>
<debug>$page_width=488976</debug>
<W>217166</W>
<D>597806</D>

<case3/>

<d>2420</d>
<width_min>192940</width_min>
<e>879.1175063482133</e>
<column divider="100">193819</column>

<d>595386</d>
<width_min>78870</width_min>
<e>216286.8824936518</e>
<column divider="100">295157</column>
...
<fo:table-column column-width="proportional-column-width(1938)"/>
<fo:table-column column-width="proportional-column-width(2952)"/>
```

![image](https://user-images.githubusercontent.com/5559926/167911427-87a47077-d6fa-4a7d-b91e-55e26e135470.png)

Will investigate where is issue.

**Comment by Intelligent2013** [comment]
1st issue fixed:
![image](https://user-images.githubusercontent.com/5559926/167930731-1fa28965-600e-4898-b7c9-1d63641c0526.png)

but padding property should be applied to tables for definition lists, or add middle columns.

**Comment by Intelligent2013** [comment]
Issue again - wrong column width:
![image](https://user-images.githubusercontent.com/5559926/168325269-743b1850-6022-44e8-a8dc-6c1307a5883b.png)

**Comment by Intelligent2013** [comment]
Issue fixed:
![image](https://user-images.githubusercontent.com/5559926/168378033-bc3e6ec4-4069-431e-bd0e-12da9f0fe4eb.png)
(text from the table cell should be aligned to the left before width calculation)

**Comment by Intelligent2013** [comment]
Wrong width for definition list in the note (left indent):
![image](https://user-images.githubusercontent.com/5559926/168378204-51bdce9d-4c72-4010-87d0-70b9f7526e8f.png)

**Comment by Intelligent2013** [comment]
XSL-FO contains block with id `id="table_if_start__4e456d6c-93ec-15c8-1aee-d2469920b647"` to determine start of the table:
```xml
<fo:inline margin-top="8pt" margin-bottom="12pt">The schemas referenced above can be found in the following parts of <fo:inline keep-together.within-line="always">ISO 10303</fo:inline>:</fo:inline>
<fo:block-container margin-left="10mm">
	<fo:block-container margin-left="0mm" margin-right="0mm">
		<fo:block margin-top="12pt">
			<fo:block>
				<fo:block font-size="1pt" keep-with-next="always" id="table_if_start__4e456d6c-93ec-15c8-1aee-d2469920b647"> </fo:block>
				<fo:table table-layout="fixed" width="95%" wrap-option="no-wrap">
					<fo:table-body>
						<fo:table-row number-columns-spanned="2">
							<fo:table-cell border="0.1pt solid black" text-align="left">
								<fo:block margin-top="0pt" id="table_if__4e456d6c-93ec-15c8-1aee-d2469920b647_1_1_p_1">
									<fo:inline font-weight="bold">
										<fo:inline keep-together.within-line="always">analysis_schema</fo:inline>
```

but IF doesn't:
```xml
<text x="41877" y="8775" dp="54 Z50 -140 Z7 -120 Z7 -140 Z35 -140 Z3 -190 Z39 -120 Z51 -120 Z15 -40 0" foi:struct-ref="1cf">The schemas referenced above can be found in the following parts of </text>
<text x="334857" y="8775" foi:struct-ref="1d0">ISO 10303</text>
<text x="379437" y="8775" foi:struct-ref="1d1">:</text>
<viewport transform="translate(28346,12000)" width="460630" height="400400">
	<viewport width="460630" height="400400">
		<g transform="translate(50,13150)">
			<border-rect x="-100" y="0" width="437698" height="12200" top="(solid,#000000,100,collapse-outer)" bottom="(solid,#000000,100,collapse-inner)" left="(solid,#000000,100,collapse-outer)" right="(solid,#000000,100,collapse-outer)" inner-background-color="#ffffff"/>
			<font weight="700"/>
			<id name="table_if__4e456d6c-93ec-15c8-1aee-d2469920b647_1_1_p_1"/>
			<text x="0" y="8875" dp="5 Z14 -260 Z3 -150 0" foi:struct-ref="1d4">analysis_schema</text>
```
It seems Apache FOP removes this block by some reason... Puzzle again...

**Comment by Intelligent2013** [comment]
Issue fixed:
![image](https://user-images.githubusercontent.com/5559926/168421831-1f84ce3c-1dfc-4340-8c28-ed99161581bd.png)

**Comment by Intelligent2013** [comment]
Done.

New algorithm for calculating table columns width available in mn2pdf v1.44 (https://github.com/metanorma/mn2pdf/releases/tag/v1.44) and ISO xslt https://github.com/metanorma/metanorma-iso/blob/main/lib/isodoc/iso/iso.international-standard.xsl
